src/main/java/mods/eln/integration/waila/GhostNodeWailaProvider.kt

```diff
@@ -0,0 +1,82 @@
+package mods.eln.integration.waila
+
+import cpw.mods.fml.common.Optional
+import mcp.mobius.waila.api.IWailaConfigHandler
+import mcp.mobius.waila.api.IWailaDataAccessor
+import mcp.mobius.waila.api.IWailaDataProvider
+import mods.eln.Eln
+import mods.eln.misc.Coordonate
+import mods.eln.packets.GhostNodeWailaRequestPacket
+import net.minecraft.client.Minecraft
+import net.minecraft.entity.player.EntityPlayerMP
+import net.minecraft.item.ItemStack
+import net.minecraft.nbt.NBTTagCompound
+import net.minecraft.tileentity.TileEntity
+import net.minecraft.world.World
+
+@Optional.Interface(iface = "mcp.mobius.waila.api.IWailaDataProvider", modid = "Waila")
+class GhostNodeWailaProvider(private val transparentNodeHandler: TransparentNodeHandler): IWailaDataProvider {
+    companion object {
+        var updateTime = Minecraft.getSystemTime()
+    }
+
+    private fun getGhostData(coord: Coordonate): GhostNodeWailaData? {
+        var ghostData: GhostNodeWailaData? = null
+
+        if (Minecraft.getSystemTime() - GhostNodeWailaProvider.updateTime > 2000) {
```

Comment:
Rather than building your own cache implementation, prefer to use Guava's. API docs here: http://google.github.io/guava/releases/20.0/api/docs/, or more concretely here: http://google.github.io/guava/releases/20.0/api/docs/com/google/common/cache/Cache.html

The `get` method would automatically load the key if nonexistent, which is bad (it'd block), but you can combine `getIfPresent` and `refresh` (from LoadingCache) to get the desired behaviour.

You're already using it elsewhere.